test/unit/core/type_traits/type_traits_iterator_test.cpp

```diff
@@ -0,0 +1,300 @@
+// Copyright (c) 2006-2020, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2020, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+#include <gtest/gtest.h>
+
+#include <seqan3/std/concepts>
+#include <forward_list>
+#include <seqan3/std/iterator>
+#include <list>
+#include <seqan3/std/ranges>
+#include <vector>
+
+#include <seqan3/core/type_traits/iterator.hpp>
+
+template <typename iterator_t>
+SEQAN3_CONCEPT iterator_traits_has_iterator_category = requires()
+{
+    typename std::iterator_traits<iterator_t>::iterator_category;
+};
+
+template <typename iterator_t>
+SEQAN3_CONCEPT has_iterator_category_tag_t = requires()
+{
+    typename seqan3::detail::iterator_category_tag_t<iterator_t>;
+};
+
+#if SEQAN3_WORKAROUND_GCC_96070
+template <typename base_t>
+struct my_iterator : base_t
+{
+    using difference_type = std::iter_difference_t<base_t>;
+    using value_type = std::iter_value_t<base_t>;
+    using reference = std::iter_reference_t<base_t>;
+    using pointer = void;
+
+    using iterator_category = seqan3::detail::iterator_category_tag_t<base_t>;
+    using iterator_concept = seqan3::detail::iterator_concept_tag_t<base_t>;
+};
+#else // ^^^ workaround / no workaround vvv
+// See https://github.com/seqan/product_backlog/issues/151 how to make this right, depending on how the standard will
+// resolve this issue.
+template <typename base_t>
+struct inherit_iterator_tag : public base_t
+{
+    using iterator_concept = seqan3::detail::iterator_concept_tag_t<base_t>;
+};
+
+template <has_iterator_category_tag_t base_t>
+struct inherit_iterator_tag<base_t> : public base_t
+{
+    using iterator_category = seqan3::detail::iterator_category_tag_t<base_t>;
+    using iterator_concept = seqan3::detail::iterator_concept_tag_t<base_t>;
+};
+
+template <typename base_t>
+struct my_iterator : inherit_iterator_tag<base_t>
+{
+    using difference_type = std::iter_difference_t<base_t>;
+    using value_type = std::iter_value_t<base_t>;
+    using reference = std::iter_reference_t<base_t>;
+    using pointer = void;
+};
+#endif // SEQAN3_WORKAROUND_GCC_96070
+
+#ifdef __cpp_lib_ranges // This is C++20 behaviour.
+TEST(iterator_category_tag_t, no_legacy_iterator)
+{
+    {
+        using view_t = std::ranges::basic_istream_view<char, char, std::char_traits<char>>;
+        using iterator_t = std::ranges::iterator_t<view_t>;
+#if SEQAN3_WORKAROUND_GCC_96070
+        EXPECT_TRUE((std::same_as<seqan3::detail::iterator_category_tag_t<iterator_t>, void>));
+#else // ^^^ workaround / no workaround vvv
+        EXPECT_TRUE(!has_iterator_category_tag_t<iterator_t>);
```

Comment:
It might be just a personal preference, but I like the style to say EXPECT_FALSE instead of EXPECT_TRUE on a negated expression. Because it is easier to read and to follow: `I expect that the expression evaluates to false` rather than `I expect the negation of the expression evaluates to true`.